src/Utilities/src/MoistThermodynamics.jl

```diff
@@ -164,44 +245,203 @@ end
+    p_vs = eval(saturation_vapor_pressure_function)(T)
+
+    return sat_shum_from_pressure(ρ, T, p_vs)
+
+end
+
+"""
+    sat_shum(T, ρ[, q_l=0, q_i=0])
+
+Compute the saturation specific humidity, given the temperature `T` and
+(moist-)air density `ρ`.
+
+If the optional liquid, and ice specific humdities `q_t` and `q_l` are given,
+the saturation specific humidity is that over a mixture of liquid and ice,
+computed in a thermodynamically consistent way from the weighted sum of the
+latent heats of the respective phase transitions (Pressel et al., JAMES, 2015).
+That is, the saturation vapor pressure and from it the saturation
+specific humidity are computed from a weighted mean of the latent heats of
+vaporization and sublimation, with the weights given by the fractions of
+condensate `q_l`/(`q_l` + `q_i`) and `q_i`/(`q_l` + `q_i`) that are liquid and
+ice, respectively.
+
+If the condensate specific humidities `q_l` and `q_i` are not given or are both
+zero, the saturation specific humidity is that over a mixture of liquid and ice,
+with the fraction of liquid given by temperature dependent `liquid_fraction(T)`
+and the fraction of ice by the complement `1 - liquid_fraction(T)`.
+"""
+function sat_shum(T, ρ, q_l=0, q_i=0)
+
+    # get phase partitioning
+    _liquid_frac = liquid_fraction(T, q_l, q_i)
+    _ice_frac    = 1 .- _liquid_frac
+
+    # effective latent heat at T_0 and effective difference in isobaric specific
+    # heats of the mixture
+    LH_0        = _liquid_frac * LH_v0 .+ _ice_frac * LH_s0
+    cp_diff     = _liquid_frac * (cp_v - cp_l) .+ _ice_frac * (cp_v - cp_i)
 
-    return 1/molmass_ratio * (1 .- q_t) .* es ./ (p .- es)
+    # saturation vapor pressure over possible mixture of liquid and ice
+    p_vs        = sat_vapor_press_generic(T, LH_0, cp_diff)
+
+    return sat_shum_from_pressure(ρ, T, p_vs)
 
 end
 
+
+"""
+    sat_shum_from_pressure(ρ, T, p_vs)
+
+Compute the saturation specific humidity, given the ambient air density `ρ`,
+temperature `T`, and the saturation vapor pressure `p_vs`.
+"""
+function sat_shum_from_pressure(ρ, T, p_vs)
+
+    _ρ_v  = p_vs ./ (R_v * T)
+    return _ρ_v ./ ρ;
+
 end
+
+"""
+    liquid_fraction(T[, q_l=0, q_i=0])
+
+Return the fraction of condensate that is liquid.
+
+If the optional input arguments `q_l` and `q_i` are not given or are zero, the
+fraction of liquid is a function that is 1 above `T_freeze` and goes to zero below
+`T_freeze`. If `q_l` or `q_i` are nonzero, the liquid fraction is computed from
+them.
+"""
+function liquid_fraction(T, q_l=0, q_i=0)
+
+q_c         = q_l .+ q_i     # condensate specific humidity
+
+# For now: Heaviside function for partitioning into liquid and ice: all liquid
+# for T > T_freeze; all ice for T <= T_freeze
+_liquid_frac = heaviside(T .- T_freeze)
+
+return ifelse.(q_c .> 0, q_l ./ q_c, _liquid_frac)
+
+end
+
+"""
+    heaviside(t)
+
+Return the Heaviside step function at `t`.
+"""
+function heaviside(t)
+   0.5 * (sign.(t) .+ 1)
```

Comment:
Do we want these functions to be type generic? In so, `0.5` would be better as `eltype(t)(1//2)`

Reply:
Made this change.